### PR TITLE
[ACR] Fix mypy errors

### PR DESCRIPTION
**Pull request by YalinLi0312**
Resolves: https://github.com/Azure/azure-sdk-for-python/issues/20640

### CODE REVIEW

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ class AnonymousACRExchangeClient(object):
     """
 
     def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constructor-parameter-credential
-        # type: (str, Dict[str, Any]) -> None
+        # type: (str, Any) -> None
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:26]: Can we use Python3 type hints?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py [hunk 2]
```diff
@@ -36,25 +36,25 @@ def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constr
...
-        self, refresh_token=None, service=None, scope=None, grant_type=TokenGrantType.PASSWORD, **kwargs
+        self, refresh_token, service, scope, grant_type, **kwargs
     ):
-        # type: (str, str, str, str, Dict[str, Any]) -> str
+        # type: (str, str, str, Union[str, TokenGrantType], Any) -> str
         access_token = self._client.authentication.exchange_acr_refresh_token_for_acr_access_token(
             service=service, scope=scope, refresh_token=refresh_token, grant_type=grant_type, **kwargs
         )
-        return access_token.access_token
+        return access_token.access_token # type: ignore
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:57]: You shouldn't need to ignore this - is `self._client.authentication.exchange_acr_refresh_token_for_acr_access_token` type hinted correctly?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py [hunk 3]
```diff
@@ -36,25 +36,25 @@ def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constr
...
+            "",
             service=parsed_challenge["service"],
             scope=parsed_challenge["scope"],
             grant_type=TokenGrantType.PASSWORD,
             **kwargs
         )
 
     def exchange_refresh_token_for_access_token(
-        self, refresh_token=None, service=None, scope=None, grant_type=TokenGrantType.PASSWORD, **kwargs
+        self, refresh_token, service, scope, grant_type, **kwargs
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:51]: Is this a public method? Or only internal? Why were all the defaults removed?
Is there any possibility that one of these wouldn't be provided?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py [hunk 4]
```diff
@@ -22,11 +22,13 @@ class ContainerRegistryChallengePolicy(HTTPPolicy):
     """Authentication policy for ACR which accepts a challenge"""
 
     def __init__(self, credential, endpoint, **kwargs):
-        # type: (TokenCredential, str, **Any) -> None
+        # type: (Optional[TokenCredential], str, **Any) -> None
         super(ContainerRegistryChallengePolicy, self).__init__()
         self._credential = credential
         if self._credential is None:
-            self._exchange_client = AnonymousACRExchangeClient(endpoint, **kwargs)
+            # pylint: disable=line-too-long
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:29]: You should just be able to add this to the one line that's too long.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py [hunk 5]
```diff
@@ -22,11 +22,13 @@ class ContainerRegistryChallengePolicy(HTTPPolicy):
     """Authentication policy for ACR which accepts a challenge"""
 
     def __init__(self, credential, endpoint, **kwargs):
-        # type: (TokenCredential, str, **Any) -> None
+        # type: (Optional[TokenCredential], str, **Any) -> None
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:25]: Is this constructor public?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 6]
```diff
@@ -143,7 +143,7 @@ def list_repository_names(self, **kwargs):
         """
         n = kwargs.pop("results_per_page", None)
         last = kwargs.pop("last", None)
-        cls = kwargs.pop("cls", None)  # type: ClsType["_models.Repositories"]
+        cls = kwargs.pop("cls", None)  # type: str
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:146]: `cls` isn't a string is it?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 7]
```diff
@@ -600,7 +600,7 @@ def update_manifest_properties(self, *args, **kwargs):
         :keyword bool can_read: Read permissions for a manifest.
         :keyword bool can_write: Write permissions for a manifest.
         :rtype: ~azure.containerregistry.ArtifactManifestProperties
-        :raises: ~azure.core.exceptions.ResourceNotFoundError
+        :raises: ~azure.core.exceptions.ResourceNotFoundError, or TypeError when any arg is in invlid type.
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:603]: You don't need to document a TypeError - this is already expected Python behaviour.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 8]
```diff
@@ -621,10 +621,16 @@ def update_manifest_properties(self, *args, **kwargs):
                 )
         """
         repository = args[0]
+        if not isinstance(repository, str):
+            raise TypeError("Parameter repository should be type of string.")
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:625]: Are these checks necessary?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 9]
```diff
@@ -621,10 +621,16 @@ def update_manifest_properties(self, *args, **kwargs):
...
         tag_or_digest = args[1]
+        if not isinstance(tag_or_digest, str):
+            raise TypeError("Parameter tag_or_digest should be type of string.")
         properties = None
         if len(args) == 3:
             properties = args[2]
+            if not isinstance(properties, ArtifactManifestProperties):
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:632]: We shouldn't check for this type.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 10]
```diff
@@ -819,21 +835,29 @@ def upload_blob(self, repository, data, **kwargs):
         :raises ValueError: If the parameter repository or data is None.
         """
         try:
-            _, _, start_upload_response_headers = self._client.container_registry_blob.start_upload(
+            _, _, start_upload_response_headers = self._client.container_registry_blob.start_upload( # type: ignore
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:838]: What were the errors here before you had to ignore them?

Comment (annatisch) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:838]: There are usually better ways to fix this than simply ignoring everything.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py [hunk 11]
```diff
@@ -143,15 +144,6 @@ def fully_qualified_reference(self):
             _strip_alg(self._digest)
         )
 
-    def _to_generated(self):
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py:146]: We never send this object?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py [hunk 12]
```diff
@@ -36,25 +36,25 @@ def __init__(self, endpoint, **kwargs):  # pylint: disable=missing-client-constr
         )
 
     def get_acr_access_token(self, challenge, **kwargs):
-        # type: (str, Dict[str, Any]) -> str
+        # type: (str, Any) -> str
         parsed_challenge = _parse_challenge(challenge)
         parsed_challenge["grant_type"] = TokenGrantType.PASSWORD
         return self.exchange_refresh_token_for_access_token(
-            None,
+            "",
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:43]: This seems suspicious - what was the reason?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py [hunk 13]
```diff
@@ -612,7 +615,7 @@ def update_manifest_properties(
     ) -> "ArtifactManifestProperties":
         ...
 
-    @distributed_trace_async
+    @distributed_trace_async # type: ignore
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py:618]: This shouldn't be necessary?

File: sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_hello_world_async.py [hunk 14]
```diff
@@ -48,7 +48,7 @@ async def basic_sample(self):
                         print(tag.digest)
 
                     # [START delete_repository]
-                    await client.delete_repository(repository_name, "hello-world")
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_hello_world_async.py:51]: Was this sample previously broken?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py (hunk 1 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:26]: To make this PR easier to review, I didn't include the hints change. I'll create a separate for it.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py (hunk 11 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_models.py:146]: This is a duplicate one.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py (hunk 12 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:43]: The value will pass to self._client.authentication.exchange_acr_refresh_token_for_acr_access_token() which is required for this parameter. I checked other languages, like in .Net is using empty string.

File: sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_hello_world_async.py (hunk 14 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_hello_world_async.py:51]: delete_repository() only accept one positional value but we provide two here, it should broken.
But, YES- all samples, not only this one, are broken as lack of resources, they all need some images uploaded in advance. I have updated this step in the samples PR.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py (hunk 5 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:25]: It is public. The hint is not correct, the credential could be None, we'll create `AnonymousACRExchangeClient` in this case.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py (hunk 3 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_anonymous_exchange_client.py:51]: - It's internal used for now. 
- Will get `Bad request` when service or scope is None.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py (hunk 4 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py]: Updated

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 6 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py]: Have removed the type hint.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 7 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py]: Updated

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 8 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py]: Have removed the type checks.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 9 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py]: Have removed.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 10 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py]: Have changed to just return headers.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py (hunk 13 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py]: Have removed

**Comment by YalinLi0312** [comment]
Updates: have addressed most of the comments except: 
1. pass `None` or `empty string` as in .Net for refresh_token while getting access token in anonymous access? 
2. should refresh_token be optional in rest API? 

We need to investigate them and will confirm with service team. cc: @annelo-msft

**Comment by annelo-msft** [comment]
> We need to investigate them and will confirm with service team.

Yes, I plan to confirm what we're seeing today, and then let's discuss with the feature crew on our Teams thread prior to reaching out to the service team.  Will post what I learn here.

**Comment by annelo-msft** [comment]
@YalinLi0312, here is [the line of code](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L130) where .NET passes an empty string when obtaining a token for anonymous access.

There is not a place where we would pass a null value to this method, because we either pass `string.Empty` as shown here, or we pass the acrRefreshToken obtained from a service call or the token cache as shown [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L115).

I found the guidance we received from the service team for obtaining anonymous access, and they shared:

> You can either use GET /oauth2/token without any auth credentials or POST /oauth2/token with `grant_type` as `password` with empty value to get a token for anonymous access.

Given this, I would recommend passing an empty value here rather than `None`, to comply with their recommendation.  If you do this, will it bypass the error you are seeing from your tools?

**Comment by YalinLi0312** [comment]
> @YalinLi0312, here is [the line of code](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L130) where .NET passes an empty string when obtaining a token for anonymous access.
> 
> There is not a place where we would pass a null value to this method, because we either pass `string.Empty` as shown here, or we pass the acrRefreshToken obtained from a service call or the token cache as shown [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Authentication/ContainerRegistryChallengeAuthenticationPolicy.cs#L115).
> 
> I found the guidance we received from the service team for obtaining anonymous access, and they shared:
> 
> > You can either use GET /oauth2/token without any auth credentials or POST /oauth2/token with `grant_type` as `password` with empty value to get a token for anonymous access.
> 
> Given this, I would recommend passing an empty value here rather than `None`, to comply with their recommendation. If you do this, will it bypass the error you are seeing from your tools?

Thanks @annelo-msft! It no longer has errors in anonymous exchange client, but the error is still there in exchange client - `refresh_token` got by calling [this ](https://github.com/Azure/azure-sdk-for-python/blob/0e3da3968eda15796cb27f80b732e50c5826095d/sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py#L77)function could be _None_,  while the expected `refresh_token` passing to [exchange_acr_refresh_token_for_acr_access_token()](https://github.com/Azure/azure-sdk-for-python/blob/0e3da3968eda15796cb27f80b732e50c5826095d/sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py#L89) is expected to be a string. Do you know how to react when refresh_token is None?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py [hunk 15]
```diff
@@ -77,7 +77,8 @@ def on_challenge(self, request, response, challenge):
         # pylint:disable=unused-argument,no-self-use
 
         access_token = self._exchange_client.get_acr_access_token(challenge)
-        request.http_request.headers["Authorization"] = "Bearer " + access_token
+        if access_token is not None:
+            request.http_request.headers["Authorization"] = "Bearer " + access_token
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py:81]: This is a good bug fix :)

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_anonymous_exchange_client.py [hunk 16]
```diff
@@ -36,25 +36,20 @@ def __init__(  # pylint: disable=missing-client-constructor-parameter-credential
             **kwargs
         )
 
-    async def get_acr_access_token(self, challenge: str, **kwargs: Dict[str, Any]) -> str:
+    async def get_acr_access_token(self, challenge: str, **kwargs: Any) -> Optional[str]:
         parsed_challenge = _parse_challenge(challenge)
         parsed_challenge["grant_type"] = TokenGrantType.PASSWORD
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_anonymous_exchange_client.py:41]: We no longer need this line?

File: sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_delete_images_async.py [hunk 17]
```diff
@@ -41,7 +48,7 @@ def __init__(self):
     async def delete_images(self):
         load_registry()
         # Instantiate an instance of ContainerRegistryClient
-        async with ContainerRegistryClient(self.endpoint, self.credential, audience=self.audience) as client:
+        async with ContainerRegistryClient(self.endpoint, self.credential, audience=self.audience) as client: # type: ignore[arg-type]
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_delete_images_async.py:51]: What were these `ignore` statements in the samples for?

**Review by annatisch — APPROVED** [review]

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_anonymous_exchange_client.py (hunk 16 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_anonymous_exchange_client.py:41]: This is a good catch, I'll remove this line, thanks!

File: sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_delete_images_async.py (hunk 17 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_delete_images_async.py:51]: I don't remember, I'll remove it and check the pipeline result.

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/samples/async_samples/sample_delete_images_async.py:51]: Interesting, the error is no longer there after remove one. I'll remove the rest.

**Pull request merged**